paasta_tools/deployd/watchers.py

```diff
@@ -23,6 +27,59 @@ def __init__(self, inbox_q, cluster):
         self.is_ready = False
 
 
+class AutoscalerWatcher(PaastaWatcher):
+
+    def __init__(self, inbox_q, cluster, **kwargs):
+        super(AutoscalerWatcher, self).__init__(inbox_q, cluster)
+        self.zk = kwargs.pop('zookeeper_client')
+        self.watchers = {}
+
+    def watch_folder(self, path):
+        """recursive nonsense"""
+        if "autoscaling.lock" in path:
+            return
+        self.log.debug("Adding watch on {}".format(path))
+        watcher = ChildrenWatch(self.zk, path, func=self.process_folder_event, send_event=True)
+        self.watchers[path] = watcher
+        children = watcher._prior_children
+        if children and ('instances' in children):
+            self.watch_node("{}/instances".format(path))
+        elif children:
+            for child in children:
+                self.watch_folder("{}/{}".format(path, child))
+
+    def watch_node(self, path):
+        self.log.debug("Adding watch on {}".format(path))
+        DataWatch(self.zk, path, func=self.process_node_event, send_event=True)
+
+    def process_node_event(self, data, stat, event):
+        self.log.debug("Node change: {}".format(event))
+        if event and (event.type == EventType.CREATED or event.type == EventType.CHANGED):
+            service, instance = event.path.split('/')[-3:-1]
+            self.log.info("Number of instances changed or autoscaling enabled for first time"
+                          " for {}.{}".format(service, instance))
+            service_instance = ServiceInstance(service=service,
```

Comment:
Hmm, I guess this could go wrong. I would think the "path" might be okay though, surely ZK isn't going to return garbage as the name of a path? Anyway, I guess we'll find out!